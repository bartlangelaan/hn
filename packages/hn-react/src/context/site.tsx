import { Site } from 'hn';
import * as React from 'react';
import globalSite from '../utils/site';

// Create a new context.
const HnContext = React.createContext(globalSite);
const { Consumer, Provider } = HnContext;

// The site provider is the same as the 'Provider', but changes 'site' to
// 'value'.
export const SiteProvider = ({
  site,
  children,
}: {
  site: Site;
  children: React.ReactNode;
}) => <Provider value={site} children={children} />;

// The consumer is exported as both Site and SiteConsumer. The SiteConsumer can
// be used if also the Site from 'hn' is imported, so the names don't overlap.
export { Consumer as Site, Consumer as SiteConsumer };

// There is also a HOC withSite that adds the site to the props of the
// component. Source of the HOC for Typescript:
// https://stackoverflow.com/a/50613946/1907875
export interface InjectedSiteProps {
  site: Site;
}

export function withSite<P extends InjectedSiteProps>(
  Component: React.ComponentType<P>,
) {
  const WrappedComponent: any = Component;
  return function ComponentWithSite(
    props: Pick<P, Exclude<keyof P, keyof InjectedSiteProps>>,
  ) {
    return (
      <Consumer>{site => <WrappedComponent {...props} site={site} />}</Consumer>
    );
  };
}

export function useHnSite(): Site {
  // @ts-ignore
  return React.useContext(HnContext);
}
