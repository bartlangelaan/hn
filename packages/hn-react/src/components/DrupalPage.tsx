import { Site } from 'hn';
import * as PropTypes from 'prop-types';
import * as React from 'react';
import { SiteConsumer } from '../context/site';
import EntityMapper, {
  EntityMapper as InnerEntityMapper,
  EntityMapperPropsMapper,
} from './EntityMapper';

interface DrupalPageState {
  error: typeof Error | null;
}

class DrupalPage extends React.Component<
  DrupalPageInnerProps,
  DrupalPageState
> {
  state: DrupalPageState = { error: null };
  lastAssuredProps: DrupalPageInnerProps | null = null;

  async bootstrap() {
    const { uuid } = await DrupalPage.assureData(this.props);

    return InnerEntityMapper.assureComponent({ ...this.props, uuid });
  }

  /**
   * Checks if we can render sychronously from a set of props.
   * If the data isn't available yet, or if an async mapper is used of which the
   * component didn't load yet, we return false. Otherwise true.
   */
  canRenderFromProps(props: DrupalPageInnerProps) {
    const uuid = props.site.getUuid(props.url);

    if (!uuid) return false;

    return !!InnerEntityMapper.getComponentFromMapper({ ...props, uuid });
  }

  lastRequest: symbol | undefined;

  componentDidMount() {
    this.componentDidUpdate();
  }

  componentDidUpdate() {
    const lastRequest = (this.lastRequest = Symbol());

    // If we can render from the current set of props, we need to change lastAssuredProps
    // to the current props. If the props change to something else that we can't render
    // from yet, we can always fallback to these props.
    if (this.canRenderFromProps(this.props)) {
      this.lastAssuredProps = this.props;
      return;
    }

    // If we can't render from props yet, we need to fetch our data and assure the mapper
    // is ready.
    (async () => {
      // First, make sure we load the url into the site cache.
      const { uuid } = await DrupalPage.assureData(this.props);

      // If in the meantime the props updated, stop.
      if (lastRequest !== this.lastRequest) return;

      // Make sure the mapper is ready by loading the async component into the mapper
      // cache.
      await InnerEntityMapper.assureComponent({
        ...this.props,
        uuid,
      });

      // If in the meantime the props updated, stop.
      if (lastRequest !== this.lastRequest) return;

      // This is in theory an unnessecary check; when we've reached this point, we should
      // be able to render from these props. However, if something did go wrong somehow,
      // we don't want to force an update. If we do that, we run into an forever running
      // loop.

      // istanbul ignore next
      if (!this.canRenderFromProps(this.props)) {
        throw Error("The data can't be loaded from the DrupalPage component.");
      }

      // Now that these props are ready, we save them for later use.
      this.lastAssuredProps = this.props;

      // We need to do a force update, because we can now synchronously load them in the
      // render function.
      this.forceUpdate();
    })().catch(error => {
      // If an error occurs in the data fetching process, we want to throw that error during
      // the render. This way it can be catched by an ErrorBoundry.
      this.setState({ error });
    });
  }

  componentWillUnmount() {
    this.lastRequest = undefined;
  }

  /**
   * This makes sure the data for this url is ready to be rendered.
   */
  static async assureData({ url, site }: { url: string; site: Site }) {
    const uuid = await site.getPage(url);

    return { uuid, pageUuid: uuid };
  }

  render() {
    if (this.state.error) {
      throw this.state.error;
    }

    const canRenderFromProps = this.canRenderFromProps(this.props);

    // If we can't render from props, and we don't want to render while loading data,
    // we return null here. That unmounts the component from the mapper, and the
    // Layout component.
    if (!canRenderFromProps && !this.props.renderWhileLoadingData) {
      return null;
    }

    // If we can render based on the props, we do.
    // Otherwise, we get the last known loaded props from the component.
    const data = canRenderFromProps ? this.props : this.lastAssuredProps;

    // Is there isn't any data available to render (also not from before), we can only
    // render the layout.
    if (!data) {
      const PropLayout = this.props.layout as React.ComponentType<any>;
      if (this.props.layout) {
        return (
          <PropLayout
            loadingData
            url={null}
            page={null}
            {...this.props.layoutProps}
          />
        );
      }

      // If there isn't a layout, we render nothing.
      return null;
    }

    const uuid = this.props.site.getUuid(data.url);

    const entityMapper = (
      <EntityMapper
        mapper={data.mapper as any}
        uuid={uuid}
        asyncMapper={data.asyncMapper as any}
        entityProps={{
          ...data.pageProps,
          page: data.site.getData(uuid),
        }}
      />
    );

    const Layout = this.props.layout;
    if (!Layout) {
      return entityMapper;
    }

    return (
      <Layout
        loadingData={!canRenderFromProps}
        url={data.url}
        page={data.site.getData(uuid)}
        {...data.layoutProps}
      >
        {entityMapper}
      </Layout>
    );
  }

  static propTypes = {
    asyncMapper: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
    ]),
    layout: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
    layoutProps: PropTypes.shape({}),
    mapper: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
    pageProps: PropTypes.shape({}),
    renderWhileLoadingData: PropTypes.bool,
    url: PropTypes.string.isRequired,
  };

  static defaultProps = {
    layout: undefined,
    layoutProps: {},
    pageProps: undefined,
    renderWhileLoadingData: false,
  };
}

export interface DrupalPageOwnProps {
  layout?: React.ReactType;
  layoutProps?: object;
  pageProps?: object;
  renderWhileLoadingData?: boolean;
  url: string;
}

type DrupalPageInnerProps = DrupalPageProps & { site: Site };
type DrupalPageProps = DrupalPageOwnProps & EntityMapperPropsMapper;

const DrupalPageWrapper = React.forwardRef<DrupalPage, DrupalPageProps>(
  (props, ref) => (
    <SiteConsumer>
      {site => <DrupalPage {...props} site={site} ref={ref} />}
    </SiteConsumer>
  ),
);

export const assureData = DrupalPage.assureData;

export default DrupalPageWrapper;
