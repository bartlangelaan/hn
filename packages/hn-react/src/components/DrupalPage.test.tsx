import React from 'react';
import renderer from 'react-test-renderer';
import site from '../utils/site';
import { asyncMapper, mapper, uuid, uuid2 } from '../utils/tests';
import waitForHnData from '../utils/waitForHnData';
import DrupalPage from './DrupalPage';

jest.mock('../utils/site', () => {
  return require('../utils/tests').mockSite();
});

jest.mock('util-deprecate', () => jest.fn(func => func));

describe('DrupalPage', async () => {
  test('with required props', async () => {
    const component = <DrupalPage mapper={mapper} url={'/'} />;

    expect(renderer.create(component).toJSON()).toMatchSnapshot();

    expect(
      renderer.create(await waitForHnData(component)).toJSON(),
    ).toMatchSnapshot();
  });

  test('with all props', async () => {
    const component = (
      <DrupalPage
        mapper={mapper}
        url={'/'}
        layout={'div'}
        layoutProps={{ testLayoutProp: true }}
        renderWhileLoadingData
        pageProps={{ testPageProp: true }}
      />
    );

    expect(renderer.create(component).toJSON()).toMatchSnapshot();

    expect(
      renderer.create(await waitForHnData(component)).toJSON(),
    ).toMatchSnapshot();
  });

  test('with asyncMapper', async () => {
    const component = (
      <DrupalPage
        asyncMapper={asyncMapper}
        url={'/'}
        layoutProps={{ testLayoutProp: true }}
        renderWhileLoadingData
        pageProps={{ testPageProp: true }}
      />
    );

    expect(renderer.create(component).toJSON()).toMatchSnapshot();

    expect(
      renderer.create(await waitForHnData(component)).toJSON(),
    ).toMatchSnapshot();
  });

  test('with asyncMapper as boolean (deprecated)', async () => {
    const component = (
      <DrupalPage
        asyncMapper
        mapper={asyncMapper}
        url={'/'}
        layoutProps={{ testLayoutProp: true }}
        renderWhileLoadingData
        pageProps={{ testPageProp: true }}
      />
    );

    expect(renderer.create(component).toJSON()).toMatchSnapshot();

    expect(
      renderer.create(await waitForHnData(component)).toJSON(),
    ).toMatchSnapshot();
  });

  test('when getPage fails', async () => {
    const component = <DrupalPage mapper={{}} url={'/newUrl'} />;

    const getPage = (site.getPage as any) as jest.Mock<any>;
    getPage.mockImplementationOnce(() => Promise.resolve('500'));

    const rendererEntry = renderer.create(component);

    await new Promise(r => process.nextTick(r));

    expect(rendererEntry.toJSON()).toBe(null);

    expect(getPage).toHaveBeenCalledTimes(1);
  });

  test('changing props', async () => {
    // Render entity 1 as usual.
    const rendererEntry = renderer.create(
      <DrupalPage mapper={mapper} url={'/'} />,
    );
    await new Promise(r => process.nextTick(r));

    expect(rendererEntry.toJSON()).toMatchSnapshot();

    // Intercept the next site.getPage call.
    let resolveGetPage;
    const getPage = (site.getPage as any) as jest.Mock<any>;
    getPage.mockImplementationOnce(
      () => new Promise(r => (resolveGetPage = r)),
    );

    // Change the url.
    rendererEntry.update(<DrupalPage mapper={mapper} url={'/new-url'} />);
    await new Promise(r => process.nextTick(r));

    // Now null should be rendered, as long as the new data is not available.
    expect(rendererEntry.toJSON()).toEqual(null);

    // Make new data available.
    resolveGetPage(uuid2);
    await new Promise(r => process.nextTick(r));

    // Now entity 2 should be rendered.
    const entity2Result = rendererEntry.toJSON();
    expect(entity2Result).toMatchSnapshot();

    // Intercept the next site.getPage call.
    let resolveGetPage2;
    getPage.mockImplementationOnce(
      () => new Promise(r => (resolveGetPage2 = r)),
    );

    // Start the loading of a new url.
    rendererEntry.update(
      <DrupalPage
        mapper={mapper}
        url={'/another-new-url'}
        renderWhileLoadingData
      />,
    );
    await new Promise(r => process.nextTick(r));

    // The component should't have changed.
    expect(rendererEntry.toJSON()).toEqual(entity2Result);

    // Add a layout component.
    rendererEntry.update(
      <DrupalPage
        mapper={mapper}
        url={'/another-new-url'}
        renderWhileLoadingData
        layout="div"
      />,
    );

    // The new layout component should have the loadingData prop set to true.
    const child = rendererEntry.root.children[0] as renderer.ReactTestInstance;
    expect(child.props.loadingData).toBe(true);

    // Intercept the next set.getPage call.
    let resolveGetPage3;
    getPage.mockImplementationOnce(
      () => new Promise(r => (resolveGetPage3 = r)),
    );

    // Load a new page (while the other page hasn't finished loading yet). Remove the layout.
    rendererEntry.update(
      <DrupalPage mapper={mapper} url={'/fresh-url'} renderWhileLoadingData />,
    );
    await new Promise(r => process.nextTick(r));

    // The component now should be the same as when before we started loading the new page.
    expect(rendererEntry.toJSON()).toEqual(entity2Result);

    // We resolve the first page. This shouldn't do anything, because it's not the latest url.
    resolveGetPage2(uuid);
    await new Promise(r => process.nextTick(r));
    expect(rendererEntry.toJSON()).toEqual(entity2Result);

    // Now we resolve the latest page. This should render entity 1.
    resolveGetPage3(uuid);
    await new Promise(r => process.nextTick(r));
    expect(rendererEntry.toJSON()).toMatchSnapshot();
  });
});
