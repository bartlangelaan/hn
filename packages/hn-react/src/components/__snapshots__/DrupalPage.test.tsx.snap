// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrupalPage changing props 1`] = `
<div
  bundle="unique_type_1__unique_bundle_1"
  className="MyCustomMapperComponent"
  entity={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  page={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  paragraph={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
/>
`;

exports[`DrupalPage changing props 2`] = `
<div
  bundle="unique_type_2__unique_bundle_2"
  className="MyCustomMapperComponentWithHOC"
  entity={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_2",
          "type": "unique_type_2",
        },
      },
    }
  }
  hocProps={true}
  page={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_2",
          "type": "unique_type_2",
        },
      },
    }
  }
  paragraph={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_2",
          "type": "unique_type_2",
        },
      },
    }
  }
/>
`;

exports[`DrupalPage changing props 3`] = `
<div
  bundle="unique_type_1__unique_bundle_1"
  className="MyCustomMapperComponent"
  entity={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  page={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  paragraph={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
/>
`;

exports[`DrupalPage with all props 1`] = `
<div
  loadingData={true}
  page={null}
  testLayoutProp={true}
  url={null}
/>
`;

exports[`DrupalPage with all props 2`] = `
<div
  loadingData={false}
  page={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  testLayoutProp={true}
  url="/"
>
  <div
    bundle="unique_type_1__unique_bundle_1"
    className="MyCustomMapperComponent"
    entity={
      Object {
        "__hn": Object {
          "entity": Object {
            "bundle": "unique_bundle_1",
            "type": "unique_type_1",
          },
        },
      }
    }
    page={
      Object {
        "__hn": Object {
          "entity": Object {
            "bundle": "unique_bundle_1",
            "type": "unique_type_1",
          },
        },
      }
    }
    paragraph={
      Object {
        "__hn": Object {
          "entity": Object {
            "bundle": "unique_bundle_1",
            "type": "unique_type_1",
          },
        },
      }
    }
    testPageProp={true}
  />
</div>
`;

exports[`DrupalPage with asyncMapper 1`] = `null`;

exports[`DrupalPage with asyncMapper 2`] = `
<div
  bundle="unique_type_1__unique_bundle_1"
  className="MyCustomMapperComponent"
  entity={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  page={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  paragraph={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  testPageProp={true}
/>
`;

exports[`DrupalPage with asyncMapper as boolean (deprecated) 1`] = `null`;

exports[`DrupalPage with asyncMapper as boolean (deprecated) 2`] = `
<div
  bundle="unique_type_1__unique_bundle_1"
  className="MyCustomMapperComponent"
  entity={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  page={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  paragraph={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  testPageProp={true}
/>
`;

exports[`DrupalPage with required props 1`] = `
<div
  bundle="unique_type_1__unique_bundle_1"
  className="MyCustomMapperComponent"
  entity={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  page={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  paragraph={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
/>
`;
